### PR TITLE
Add documentation for authorization on the REST layer

### LINKED ISSUE #4381: [DOC] Add doc for Authorization in REST Layer
**What do you want to do?**
 
- [ ] Request a change to existing documentation
- [x] Add new documentation
- [ ] Report a technical problem with the documentation
- [ ] Other 

**Tell us about your request.** Provide a summary of the request and all versions that are affected.

## Description of the Feature

The Security plugin has recently introduced a new feature. This addition enhances the plugin's capabilities by enabling user authorization for requested REST routes. Let's delve into the details of how this feature functions.

**NOTE: This feature sits on top of existing authorization check at Transport Layer. No changes were made at Transport Layer.**

During the registration of a new route, the plugin requires a unique name to be assigned to each route. This ensures that no duplicate names are used during the registration process. Additionally, the plugin provides the option to register a set of "legacy" cluster permissions for each route. These permissions, such as `cluster:admin/transport/`, serve as the criteria against which the route is evaluated when a user attempts to access it.

## Functionality of the Feature

The authorization process for this feature operates as follows: when a user initiates a REST request, their roles are examined, and each permission associated with the user is evaluated. The evaluation is conducted to determine if either of the following conditions is met: a match with the unique name assigned to the route or a match with any of the legacy actions defined during the route's registration.

For instance, let's consider a scenario where a user is assigned to role A. Here is an example configuration in YAML format:

```yml
abcplugin_read_access:
  reserved: true
  cluster_permissions:
    - 'cluster:admin/opensearch/abcplugin/route/get'
    - 'cluster:admin/opensearch/abcplugin/route/put'
```

Suppose the plugin has two registered routes under the `abcplugin` namespace as instance of [NamedRoute](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/NamedRoute.java):

1. `/_plugins/_abcplugin/route/get` registered with the unique name `routeGet` and associated with the legacy transport action `cluster:admin/opensearch/abcplugin/route/get`.
2. `/_plugins/_abcplugin/route/delete` registered with the unique name `routeDelete` and associated with the legacy transport action `cluster:admin/opensearch/abcplugin/route/delete`.

In this scenario, when the user makes a REST API call to `/_plugins/_abcplugin/route/get`, they will be granted authorization. The request flow will proceed as normal. 
However, if the same user attempts a REST call to `/_plugins/_abcplugin/route/delete`, they will be denied authorization and receive an error message.

## Goal of the documentation

1. It should help plugin developers point to the correct resources, for registering named routes
2. It should help end user define correct permission actions in the roles.
 

Related issue: https://github.com/opensearch-project/security/issues/2590

### PR DESCRIPTION
**Pull request by cwillum**
### Description
To support extensions, authorization of API actions can now be carried out on the REST layer, necessarily for extensions and optionally for new and existing plugins. This PR documents details surrounding the enhancement.

### Issues Resolved
Fixes #4381 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by cwillum** [comment]
Just collecting the dots at this point. Thus the draft status.

**Comment by cwillum** [comment]
Received news that this will most likely get pushed to 2.10.

**Comment by cwillum** [comment]
@DarshitChanpura Could you have a look at this for technical accuracy, **language**, and completeness? I'm not sure if I got some of the language-concept agreement right in some cases. Thanks.

File: _security/access-control/rest-layer-authz.md
Comment (DarshitChanpura) [inline _security/access-control/rest-layer-authz.md:11]:
```suggestion
Authorization on the REST layer provides an added level of security for plugin and extension API requests by offering a mechanism for authorization checks on the REST layer. This level of security sits atop the transport layer and provides a complementary method of authorization without replacing, modifying, or in any way changing the same process on the transport layer. REST layer authorization was initially created to address the need for an authorization check for extensions, which do not communicate on the transport layer. However, the feature is also available for developers to utilize while creating future plugins to operate with OpenSearch.
```
Current plugins have not been migrated to utilize this feature. But, the feature is available for current plugins to migrate their APIs to support authorization.

File: _security/access-control/rest-layer-authz.md
Comment (DarshitChanpura) [inline _security/access-control/rest-layer-authz.md:15]:
```suggestion
The benefits to developers when using the REST layer for authorization mean that they can authorize requests at REST layer and filter out unauthorized requests. As a result, this decreases the processing burden on transport layer while allowing granular control over access to APIs.
```

We are not replacing transport authz, we are just adding a layer on top.

File: _security/access-control/rest-layer-authz.md
Comment (DarshitChanpura) [inline _security/access-control/rest-layer-authz.md:31]: This is true only for extensions. For plugins, developers will have to declare the permission names.

**Review by DarshitChanpura — CHANGES_REQUESTED** [review]
Some minor changes requested. Thank you @cwillum for putting this together.

File: _security/access-control/rest-layer-authz.md
Comment (cwillum) [reply _security/access-control/rest-layer-authz.md:31]: @DarshitChanpura (1) What exact part of "this" is true for creating a permission name for extensions? Is it this whole piece: "... the `routeNamePrefix` value in the `settings.yml` file for the resource `ad` is added to the route to complete a unique name."?
(2) Is there any guidance for declaring a permission name for plugins? Are there required elements in the name like that for a permission name for an extension? Or do you just invent anything you want?

Comment (DarshitChanpura) [reply _security/access-control/rest-layer-authz.md:31]: Apologies for confusion. By "this" I mean the `routeNamePrefix` is a setting only configured in extension. This setting doesn't exist in plugins.

There is no guidance as of now in defining permissions. But we can make a suggestion to have it something like: `<plugin/extension-name>:<route>` (i.e. `ad:detectors/get`)

Let me know if this helps.

Comment (cwillum) [reply _security/access-control/rest-layer-authz.md]: @DarshitChanpura Thanks for the clarification. I've adjusted that wording to include current understanding for extensions and plugins. I've also addressed the other two comments/suggestions. I think this should be ready for documentation team review when 2.11 comes along. I'll comment that it's on hold for that. And if there are any changes between now and then, we can always revise the current draft before the release.

**Comment by cwillum** [comment]
The work is essentially done for this PR. I considered adding the status tag "Done but waiting to merge". But it will need documentation team review and editorial review before merging for 2.11. I'm leaving it alone for now in case the Security team wants any additional changes before then.

**Review by DarshitChanpura — APPROVED** [review]
LGTM!

**Comment by cwillum** [comment]
Stepping this up to doc review.

File: _security/access-control/rest-layer-authz.md
Comment (Naarcha-AWS) [inline _security/access-control/rest-layer-authz.md:15]: I would combine this sentence with last paragraph of the first to avoid confusion between the different personas:

```suggestion
Developers, on the other hand, will need to understand the ideas behind `NamedRoute` and how the new route scheme is constructed. The benefits to developers when using the REST layer for authorization include the ability to authorize requests at the REST layer and filter out unauthorized requests. As a result, this decreases the processing burden on the transport layer while allowing granular control over access to APIs.  For detailed information about the new route scheme, see [Authorization at REST Layer for plugins](https://github.com/opensearch-project/security/blob/main/REST_AUTHZ_FOR_PLUGINS.md). 
```

File: _security/access-control/rest-layer-authz.md
Comment (Naarcha-AWS) [inline _security/access-control/rest-layer-authz.md:31]:
```suggestion
For extensions, you can create a `NamedRoute` from this by referencing the `routeNamePrefix` value in the `settings.yml` file for the resource (`ad` in this case) and adding it to the route to complete a unique name. The result is shown in the following example:
```

**Review by Naarcha-AWS — APPROVED** [review]
Two small fixes, otherwise LGTM.

File: _security/access-control/rest-layer-authz.md
Comment (cwillum) [reply _security/access-control/rest-layer-authz.md:15]: I get your point. That's a good idea. I just divided the two statements about users and developers, and left the "benefits" statement on its own, with a little rewording.

**Review by DarshitChanpura — APPROVED** [review]
LGTM. @Naarcha-AWS Would you please re-review this?

File: _security/access-control/rest-layer-authz.md
Comment (natebower) [inline _security/access-control/rest-layer-authz.md:13]:
```suggestion
For users that work with REST layer authorization, the methods of assigning roles and mapping users and roles, and the general usage of plugins and extensions, remain the same---the only additional requirement being that users become familiar with a new scheme for permissions. 
```

File: _security/access-control/rest-layer-authz.md
Comment (natebower) [inline _security/access-control/rest-layer-authz.md:15]: It's probably worth noting that up to this point, we used "on" the REST layer as opposed to "at". We should probably be consistent throughout.

File: _security/access-control/rest-layer-authz.md
Comment (natebower) [inline _security/access-control/rest-layer-authz.md:19]: "You must enable the Security plugin in order to use REST layer authorization"? (to avoid the alternate meaning of the current construction)

File: _security/access-control/rest-layer-authz.md
Comment (natebower) [inline _security/access-control/rest-layer-authz.md:27]: First sentence: "the OpenSearch Project has introduced"? Last sentence: registered "to"?

Comment (natebower) [inline _security/access-control/rest-layer-authz.md:27]:
```suggestion
To facilitate REST layer authorization, OpenSearch introduces the idea of [`NamedRoute`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/NamedRoute.java) for route registration. For developers, this standard requires a new method of registering routes that uses a unique name. While transport actions typically consist of a method name, a part, and a corresponding transport action, this new implementation requires a method name, a part, and a unique name for the route. As the name suggests, it is essential that it be unique among all plugins and extensions or, in other words, not registered for any other route.
```

File: _security/access-control/rest-layer-authz.md
Comment (natebower) [inline _security/access-control/rest-layer-authz.md:51]:
```suggestion
There is no change to the way you map users and roles with `NamedRoute`. Also, the new format for the permission is compatible with existing configurations. This section provides an example of how user and role mappings appear for legacy and `NamedRoute` configurations and how they authorize registered routes for actions.
```

File: _security/access-control/rest-layer-authz.md
Comment (natebower) [inline _security/access-control/rest-layer-authz.md:71]: Would "For a different route, such as..." work here?

**Review by natebower — COMMENTED** [review]
@cwillum Just a few comments/changes. Thanks for your advance work on this!

File: _security/access-control/rest-layer-authz.md
Comment (cwillum) [reply _security/access-control/rest-layer-authz.md:15]: RE: Authorization at REST Layer for plugins
Yeah, I can't do anything about the name of that resource. It's created by Security team developers. But point taken.

**Pull request merged**
